### ISSUE TITLE
AnimatedView

### ISSUE THREAD

**Issue by nick-thompson** [nick-thompson/blueprint#31]
The `blueprint::View` has a `setFloatBounds` method that the ShadowView calls into during a layout step. That method then delegates on to the `juce::Component`. I want to continue keeping this piece decoupled from the `juce::Component` interface, but there's a neat opportunity here because of the decoupling:

I want to create an `AnimatedView`, similiar to juce's `ComponentAnimator`, where when the `View` receives a call to `setFloatBounds`, instead of immediately calling the `juce::Component::setBounds`, start a timer at 60Hz or something to animate the bounds of the `juce::Component`. Basically on every tick of the timer nudge the current bounds closer to the target bounds until they're close enough to equal, at which point we stop the timer. How much to nudge on every frame can be configured via a settable property with easeIn, easeOut, linear, etc, options.

**Comment by JoshMarler** [comment]
Hey @nick-thompson ! 

I'm just starting to look into this as I need a horizontal Expander/Accordion component which expands and shows some content in response to a button click event etc. 

So, it seems like just wrapping `juce::ComponentAnimator` in a `blueprint::AnimateView` or similar would be reasonably straight forward. However I'm curious about your ideas and obviously would like to contribute something directly if I can rather than rolling my own interpretation of this. 

So you mention starting a timer and gradually incrementing/decrementing bounds, basically exactly what `juce::ComponentAnimator` does. What's the rational behind doing blueprint's own thing here as opposed to just wrapping `juce::ComponentAnimator` ? 

I've also been looking at the API for ReactNative's `Animator` API: https://reactnative.dev/docs/animations

Obviously this is much more comprehensive than the simple bounds/opacity animations offered by `juce::ComponentAnimator`. I wonder whether mimicking it's API is worth consideration though, it could provide a powerful framework to follow. I appreciate this all has a bit of overlap with the <Canvas> API's animation offerings. 

Be cool to get a discussion going on this one!

Cheers

**Comment by JoshMarler** [comment]
This also looks like a relevant bit of API inspiration: https://reactnative.dev/docs/layoutanimation

**Comment by nick-thompson** [comment]
Awesome, yea, happy to kick this convo off! I think this would be a really fun component to have.

I've studied the RN APIs quite a bit already as I've thought about this task. I have a couple high level thoughts and I actually also have a branch I started months ago that has this mostly working. I'll push the branch up just for discussion's sake in a minute.

So first off, yea, you're totally right, `juce::ComponentAnimator` will do fine. At some point I had a reason why it might be helpful for us to roll our own, but I guess it wasn't a very good reason because I can't remember it :)

My main thought on this: I think `AnimatedView` should provide layout animation only, and I say that because it covers many use cases and has a very simple and elegant implementation on the native side. 

React Native has this notion of an "Animated Value" which I think we could try to emulate, but this would be tricky and I would suggest that we build out the simple AnimatedView first and defer any notion of an "Animated Value" to a later, longer discussion. In particular, animation relies on some quick timer firing to crunch new numbers, and I think we will certainly want to avoid invoking javascript on each animation frame. That could just accrue so fast– multiple AnimatedValues, each invoking js on a 30Hz timer, etc. So any notion of an AnimatedValue that we have should hand off animation parameters to native where we run the timer. But then passing an AnimatedValue instance to any View component, the View would have to know when and how to ask for updates from the AnimatedValue. It gets complicated...

Anyway, for this task let's focus on layout, and it can be accomplished really easily by designing a new `AnimatedView` which is just a regular `View` instance but with a custom `AnimatedShadowView`, where the `AnimatedShadowView`, upon receipt of new layout directives, computes the new layout bounds and uses `juce::ComponentAnimator` to apply them over time instead of apply them immediately. The whole Yoga/ShadowView architecture makes this quite simple :). Let me push my branch to help the discussion, and then we can keep the conversation going here to see what else might need to be done/changed.

**Comment by nick-thompson** [comment]
Ok here we go: https://github.com/nick-thompson/blueprint/compare/feature/animatedviewscratch?expand=1

Also, I found in that code the reason why we might want to roll our own animator! See the big `TODO` inline in the code... now that we have the `CanvasView` I think this could be quite important (each time `paint()` is called you would want to make sure that the width/height you got in `onMeasure()` is accurate)

I remember also there was something buggy in here as I was working on it, but I think it's a good start and helpful for the discussion anyway

**Comment by JoshMarler** [comment]
Awesome thanks Nick. I'll pull down that branch and have a play, wondering if you could add a component listener to view instance inside flushLayout and call setFloatBounds in response to size/positions changes triggered by the animator or something. 

I'm just spitballing whilst on a walk reading this on my phone though so I'll take a concentrated look in the morning and drop you a PR if I get anything useful!

**Comment by JoshMarler** [comment]
The "Animated Value" stuff is a super interesting idea and I'd love to help investigate. It'd be so awesome for Blueprint to bring a concrete animation approach to juce, it's one area in juce's GUI offerings that is a little lacking.

**Comment by nick-thompson** [comment]
> The "Animated Value" stuff is a super interesting idea and I'd love to help investigate. It'd be so awesome for Blueprint to bring a concrete animation approach to juce, it's one area in juce's GUI offerings that is a little lacking.

Yea, totally agree! I'm highly confident in our layout animation. I'm curious what sorts of animations people would be interested in other than layout (do you have a specific case already?). I say that because my gut feeling is to add the `AnimatedView` for layout animation now, and then tackle the `AnimatedValue` super slowly afterwards to make sure we're thoughtful about it. I anticipate a generic `AnimatedValue` being pretty difficult to get right.

I'm going to update my branch in just a minute– I realized just now that the `juce::ComponentAnimator` implements a `ChangeBroadcaster`, so we can tap into that to easily fix the float bounds issue. Will report back shortly

**Comment by JoshMarler** [comment]
Yeah awesome I figured we should be able to add a listener and then dispatch the changes to setFloat kinds somehow. Nice one. Shout if you want me to take anything on that one. Have a few PRs to drop over this week. 

Totally agreed on AnimatedValue concepts. I'll get some noted together on some of the things I expect to need in future and maybe we can start a new issue just to open the discussion. 

Cheers Nick

**Comment by nick-thompson** [comment]
Awesome, thanks.

So I just set out to fix up my branch and noticed two things, curious to hear your thoughts on it:

1) So the `juce::ComponentAnimator` _does_ implement a ChangeBroadcaster, and emits a change on every animation frame, but there's no way to query the animator for what the _current_ bounds of the target component are. We can query the `view` itself in the change callback, but the unfortunate part about that is that by that point JUCE has already truncated to integer bounds (Blueprint tries hard to keep all bounds floating point until the last possible minute, otherwise you get very noticeable pixel boundary snapping during resize and such). I feel like this is a +1 for writing our own layout animator?

2) The way the `AnimatedView` is written on my branch right now, the `AnimatedView` itself will animate, but its children will not. For example:

```js
<AnimatedView>
  <View>
    <Text>Longform text........Longform text........Longform text........Longform text........</Text>
  </View>
</AnimatedView>
```

If you, for example, drastically reduce the width of the parent AnimatedView, the inner View and Text will snap to their target destinations immediately while the parent animates. You could partially solve this just by making the child `View` an `AnimatedView` as well. But with `<Text>` you'll still get the snap. The first question here is: is this sensible behavior? Or would you expect that all children of an `AnimatedView` _also_ animate into place when the parent animates? I have ideas in mind for how to implement both approaches but curious to hear from you which one feels more correct.

**Comment by JoshMarler** [comment]
Hey @nick-thompson, 

Hmm yeah, so with any juce based listener pattern there we're going to get rounding errors when trying to call `setFloatBounds`, I hear ya. Yuk. The `juce::ComponentAnimator` implementation looks like it handles a few little edge-cases etc. Maybe we can rip some of it off and benefit from jules's wisdom? Take the good old copy-paste inheritance design-pattern (My old man started coding in the 80's and he's convinced it still beats anything "OO" has to offer).

In regards to point 2, personally I thiiiiink I'd want the animation to trickle down through all of the child view nodes so that everything animates at the same rate based on the parent component's animating bounds. I'd be curious to know what the use-cases for the other "snap children" approach would be. 

Happy to have a crack at a blueprint "Animator" if you're snowed under. 

Cheers

Josh

**Comment by nick-thompson** [comment]
Yea, I think I'm with you on all fronts. I'm quite intruiged by this one, I'll have a crack at it now for a bit. If you need it sooner or if I get stuck I'll probably try to hand it off to you :)

**Comment by JoshMarler** [comment]
Yeah awesome! No worries. I shall continue with me other little PR's. Thought I might drop you a few fixes for some other open issues anyways.

**Comment by nick-thompson** [comment]
Ok check this out, this is my new approach: https://github.com/nick-thompson/blueprint/compare/feature/animation-again?expand=1

Basically, in order to recursively animate view children, we can't rely on a new shadow view. We need any old shadow view to be told "you gotta animate now." So I added a little hook to the `flushViewLayout` in the default ShadowView that checks for an `animated` prop _or_ receives a directive from its parent. Every shadow view that needs to animate then just uses a custom little interpolator and a callback hook to make sure all the state is updated correctly.

I looked into copying juce's component animator and actually ended up going a different direction, and I think it's quite neat. Let me know what you think

**Comment by JoshMarler** [comment]
@nick-thompson, This is really cool! I really love the way you've solved recursively animating the children there. 

So I have a couple initial thoughts. 

1. Would `layoutAnimated` or `animateLayout` be a better prop name for the containing `View` just to avoid confusion with any future APIs etc. ?

2. It would be nice to be able to pass some basic arguments (maybe via some `config` object) to `layoutAnimated`/`animateLayout`

i.e.

`<View animateLayout={duration: 200, interp_type: SOME_INTERPOLATION_ENUM} />`

You could allow `<View animateLayout={} />` or similar to simple apply a sensible set of defaults for the animation. 

Over all though I think this looks super clean. Nice one!

**Comment by nick-thompson** [comment]
Awesome, glad you like it.

To your feedback– yea, all of it! I like the idea of using an object as the prop value, and with a prop name that makes sense on its own or with args. For example, in React if you type `<View layoutAnimated>` it's equivalent to writing `<View layoutAnimated={true}>`. I think it might be cool to use that approach if you want the defaults, or supply an object if you want specification:

```
<View layoutAnimated />
<View layoutAnimated={{duration: 200, easing: View.Animation.QuadInOut, frameRate: 45 }} />
```

On the implementation side we just say `props.contains("layoutAnimated")` and `props["layoutAnimated"].isBool()` to make the distinction, etc.

Totally open to continuing a discussion on the best prop name. I'll go ahead in this direction in the meantime and can circle back to rename any time

**Comment by JoshMarler** [comment]
@nick-thompson, yeah awesome I agree with all of that. No strong opinion on the prop name other than `animated` feels like it could make the API a little confusing. `layoutAnimated` seems perfectly good to me.

**Comment by JoshMarler** [comment]
Hey @nick-thompson! How's this beast going? Can I assist with anything? No pressure! Just wondered if you'd hit any major stumbling blocks. I'm starting to deal with more of the animated/dynamic components in our app. Planning to start getting a list of performance hot spots together etc.

**Comment by nick-thompson** [comment]
Ah, sorry, I just haven't had a chance yet to wrap it up. I'll try to find some time as soon as I can!

**Comment by JoshMarler** [comment]
No probs! Was curious to hear if you'd hit some horrible trap as I'm poking around with similar things.

**Comment by nick-thompson** [comment]
Closing this as fixed as of #85
